### ISSUE TITLE
`renku log` to generate users resource pointing the Renku `users` resource

### ISSUE THREAD

**Issue by jachro** [SwissDataScienceCenter/renku-python#662]
As a Renku owner, I'd like Renku log to generate users/agents/creators resource triples pointing to the new GET /knowledge-graph/users/:id endpoint. The `id` path parameter is a `UUID`.

**Comment by rokroskar** [comment]
who creates the UUID? We could use the keycloak user id? This would mean we have to let the user log in to the platform from the CLI and retrieve their keycloak bearer token.

**Comment by jachro** [comment]
Yeah, the UUID would be great but you're right, it's quite challenging to get one...

What about asking the `knowledge-graph` to generate one? You simply call some search endpoint (by giving an email and/or orcid-id or whatever you have) and you'll get a UUID - either existing or generated one. I can imagine, they will be cases when you cannot access the service (when you work with Renku locally) so in such cases `renku-cli` would generate something meaningful. But once you push your commit to GitLab, the service will be there and all should be ok (in theory at least).

I guess having something central for the user resource generation would be pretty cool.

Any ideas/comments?

**Comment by rokroskar** [comment]
We should use the same ID that the oauth services are using. You can see it if you go to e.g. https://renkulab.io/jupyterhub/services/notebooks/user under `auth_state.gitlab_user.identities[0].extern_uid`

**Comment by jachro** [comment]
That would be nice but the problem I see here is users who are not our Renku users like dataset authors which we import from e.g. Zenodo. Maybe I'm a bit perfectionist here but I'd really like if all users can be identified the same way :) Of course, the oauth ID can be maybe added as another `sameAs` resource (or something similar).

**Comment by rokroskar** [comment]
If we have a `users` endpoint it would make sense that it only applies to the _users_ of the renku instance. If we want to search/access the creators of datasets, then those should be accessible through the `creators` endpoint. A `creator` can be anyone, either a `user` or someone who was added to the KG by virtue of being associated with a dataset or some other entity.

**Comment by jachro** [comment]
Yes, that's probably a better solution. It actually makes a lot of sense as a user is a Renku user, not a random person from wherever.

So in terms of the identifier, should we stick to the ID that the OAuth services are using?

**Comment by rokroskar** [comment]
I would do that, but then we need the means to log the user in via the CLI. This might be a slightly more involved process, and I'm not sure we currently have an issue for it. There was a very old PR opened once to allow us to do this, but I'm not sure what happened with it. Can you make an issue for it so we don't forget? We can always reopen the old one if needed and close the new one if it turns out it's a duplicate.

**Comment by ableuler** [comment]
yep, logging in a user in through the CLI was implemented by @ebiii a long time ago (see [this PR](https://github.com/SwissDataScienceCenter/renku-python/pull/193)). Some dust might have accumulated on top of the feature, but I think it could be reactivated.

**Comment by jachro** [comment]
I went through the PR but it's not that clear to me how it relates to changing the users resources in the triples to smth like this: http://renku/users/<OAuth-ID>. But if that works for you I can re-open it.

**Comment by rokroskar** [comment]
It implements logging in which means that we can retrieve the keycloak-issued JWT into the local client. This gives us access to the instance-global user ID.

**Comment by jachro** [comment]
So should I re-open it?

**Comment by Panaetius** [comment]
We already use UUIDs that could be resolvable in renku-python. Should this be moved to UI/gateway since that's the place where it needs to be resolved?

I don't think there's something to be done on the renku-python side for this.
